Sources/Purchasing/VerificationResult.swift

```diff
@@ -69,3 +69,29 @@ extension VerificationResult: DefaultValueProvider {
     static let defaultValue: Self = .notVerified
 
 }
+
+extension VerificationResult {
+
+    /// - Returns: the most restrictive ``VerificationResult`` based on the cached verification and
+    /// the response verification.
+    static func from(cache cachedResult: Self, response responseResult: Self) -> Self {
```

Comment:
Oh, I see that here you will return the response as not valid if a 304 is not valid. I don't think we should do this. We should act as unable to fetch. There is a legit situation for this, imagine you are in a airport wifi that intercepts and mangles requests. It should not break your cache.

I think I still have trouble with the logic for validation being in two places.

I would rather do something like:
- For validation errors (either 200, 403...), if there is cache, treat this as a 500, just return cache
- For 403 update request time in cache only if cache was checked valid and 304 was valid too (304 invalid would be covered in the above case anyway). Then return cache.
- For valid 200, update cache, return fresh data

Pseudocode:
```
# On requests:
if cache and (cacheValidationOk or SDKValidationDisabled):
    request_with_cache_etag()
else:
    request_without_etag()
```

```
def get_response_or_cache():
# On responses:
wrong_response = response_not_modified and (we_did_not_sent_etag or returned_etag_mistmatch)
  return FailedRequest  # This should not happen, we only expect not modified if we have a etag cached reponse. If the cache data is not trusted we should NOT send etag always to force a refetch (unless SDKValidationDisabled).

if failedRequest or wrong_response:
    if cache:
        return cache
    else:
        return FailedRequest

trust_response = responseValidationOk or SDKValidationDisabled

if trust_response:
    if not_modified:
        update_cache_request_time()
        # While we might trust the response if validation is disabled, we still track
        # wether the cached data is valid or not, and it will be only valid if cache
        # and requests are both valid
        update_cache_validation_to(responseValidationOk and cacheValidationOk)
        return cache
    else:
        store_new_response_in_cache()
        update_cache_validation_to(responseValidationOk)
        return response
elif cacheValidationOk:
    # If we are not trusting response, prefer valid cache always
    return cache
else:
    # We do not trust response nor cache
    if not_modified:
        return cache
    else:
        store_new_response_in_cache()
        update_cache_validation_to(validationFailed)
        return response
```

This code will handle properly the cache and validation status. In some cases this will return responses with validation failures, but it will try its best to use valid cache over a failed validation response (unless we are using the disable validation mode, in which case uses all responses valid or not (but still tracks validation state irregardless).

Above this, you can put the enforcer layer for strict mode. Having it separate simplifies the request / cache mangling layer.
```
    response = get_response_or_cache()
    if SDKValidationStrictMode and not response.isValid:
        return failedValidation
    return response
```

Reply:
It took me a full hour write what I was thinking was going to be simple pseudocode. This request handling logic is truly complicated XDDD

Perhaps it is better to sync. Ping me if you are available.

Reply:
> Oh, I see that here you will return the response as not valid if a 304 is not valid. I don't think we should do this. We should act as unable to fetch. There is a legit situation for this, imagine you are in an airport wifi that intercepts and mangles requests. It should not break your cache.

> I think I still have trouble with the logic for validation being in two places.

> I would rather do something like:

> For validation errors (either 200, 403...), if there is cache, treat this as a 500, just return cache

That's not changed. As per my other comment, that would be handled by `DeviceCache`. `HTTPClient` doesn't return a valid response if there is a 500.

> For 403 update request time in cache only if cache was checked valid and 304 was valid too (304 invalid would be covered in the above case anyway). Then return cache.
> For valid 200, update cache, return fresh data

That's how it works 👍🏻